lib/proxy/peer/client.go

```diff
@@ -214,14 +254,87 @@ func (c *clientConn) Shutdown(ctx context.Context) {
 	}
 }
 
-// Client is a peer proxy service client using grpc and tls.
+// close implements [clientConn].
+func (c *grpcClientConn) close() error {
+	return c.cc.Close()
+}
+
+// dial implements [clientConn].
+func (c *grpcClientConn) dial(
+	nodeID string,
+	src net.Addr,
+	dst net.Addr,
+	tunnelType types.TunnelType,
+) (net.Conn, error) {
+	release := c.maybeAcquire()
+	if release == nil {
+		c.metrics.reportTunnelError(errorProxyPeerTunnelRPC)
+		return nil, trace.ConnectionProblem(nil, "error starting stream: connection is shutting down")
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	context.AfterFunc(ctx, release)
+
+	stream, err := clientapi.NewProxyServiceClient(c.cc).DialNode(ctx)
+	if err != nil {
+		cancel()
+		c.metrics.reportTunnelError(errorProxyPeerTunnelRPC)
+		return nil, trace.ConnectionProblem(err, "error starting stream: %v", err)
+	}
+
+	err = stream.Send(&clientapi.Frame{
+		Message: &clientapi.Frame_DialRequest{
+			DialRequest: &clientapi.DialRequest{
+				NodeID:     nodeID,
+				TunnelType: tunnelType,
+				Source: &clientapi.NetAddr{
+					Addr:    src.String(),
+					Network: src.Network(),
+				},
+				Destination: &clientapi.NetAddr{
+					Addr:    dst.String(),
+					Network: dst.Network(),
+				},
+			},
+		},
+	})
+	if err != nil {
+		cancel()
+		return nil, trace.ConnectionProblem(err, "error sending dial frame: %v", err)
```

Comment:
I just moved the grpc dialing code (and changed the `append(errs, ...)` into `return ...`), changing it would be a bit too out of scope for this I think.